File: iteration1/src/Services/StudentCreator.java
package iteration1.src.Services;

import iteration1.src.Models.*;
import java.util.ArrayList;
import java.util.Calendar;

public class StudentCreator {
    private Semester semester;
    private Advisor advisor;
    private ArrayList<Course> takenCourses;

    public StudentCreator(Semester semester, Advisor advisor, ArrayList<Course> takenCourses) {
        this.semester = semester;
        this.advisor = advisor;
        this.takenCourses = takenCourses;
    }


    public Student createRandomStudent(int index, String name) {
        StudentID studentID = generateStudentID(index);

        TranscriptCreator tc = new TranscriptCreator();
        Transcript transcript = tc.generateTranscript(takenCourses);

        return new Student(name, studentID, semester, advisor, transcript);
    }

    // semester, index | done
    private StudentID generateStudentID(int index) {
        String department = "1501";
        String entryYear = ((Calendar.getInstance().get(Calendar.YEAR) - (int) (Math.ceil(semester.getSemesterNo() / 2.0))) % 100) + "";
        String entryOrder;
        if (index < 10) {
            entryOrder = "00" + index;
        } else if (index < 100) {
            entryOrder = "0" + index;
        } else {
            entryOrder = "" + index;
        }

        return new StudentID(department, entryYear, entryOrder);
    }


}

File: python_project/src/Services/StudentCreator.py
# fmt: off
import datetime
import math
import sys
sys.path.append("..")
from Models.Student import Student
from Models.StudentID import StudentID
from Models.Transcript import Transcript
# fmt: on


class StudentCreator:
    def __init__(
        self,
        semester,
        advisor,
        failRate,
        takenCourses: None,
        teCourses: None,
        nteCourses: None,
    ):
        self.semester = semester
        self.advisor = advisor
        self.failRate = failRate

        if takenCourses is None:
            self.takenCourses = []
        else:
            self.takenCourses = takenCourses

        if teCourses is None:
            self.teCourses = []
        else:
            self.teCourses = teCourses

        if nteCourses is None:
            self.nteCourses = []
        else:
            self.nteCourses = nteCourses

    def generateStudentID(self, index: int):
        department = "1501"
        entryYear = str(
            datetime.date.today().year - math.ceil(self.semester.semesterNo / 2)
        )[2:]
        entryOrder = str(index).zfill(3)

        return StudentID(department, entryYear, entryOrder)

    def createRandomStudent(self, index: int, name: str):
        studentID = self.generateStudentID(index)
        transcript = Transcript()
        transcript.generateTranscript(
            self.failRate,
            self.semester,
            self.takenCourses,
            self.teCourses,
            self.nteCourses,
        )

        return Student(name, studentID, self.semester, self.advisor, transcript)
